File: Data Structures & Algorithms/subsets-ii/submission-0.java
class Solution {
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> set = new ArrayList<>();
        int i = 0;
        Arrays.sort(nums);
        dfs(res, set, nums, i);
        return res;
    }

    private void dfs(List<List<Integer>> res, List<Integer> set, int[] nums, int index){
        if(index > nums.length){
            return;
        }

        if(index <= nums.length){
            res.add(new ArrayList<>(set));
        }
       
        for(int i = index; i < nums.length ; i++){
            if(i > index && nums[i] == nums[i - 1]){
                continue;
            }
            set.add(nums[i]);
            dfs(res, set, nums, i + 1);
            set.remove(set.size() - 1);
        }

    }
}

File: Data Structures & Algorithms/contains-duplicate-ii/submission-0.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        st = set()
        i = 0
        j = 0
        while j < len(nums):
            if j - i > k:
                st.remove(nums[i])
                i += 1

            if nums[j] in st:
                return True
            
            st.add(nums[j])
            j += 1

        
        return False
